File: code/code34.py
from typing import List
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        """
        l:找到第一个大于等于target的
        r:找到第一个大于target
        """
        if not nums:
            return [-1,-1]
        def binSeach(target: int) -> int:
            l = 0
            r = len(nums) # 这里故意右移一位
            while (l < r):
                mid = (l + r) // 2
                if nums[mid] < target:
                    l = mid + 1
                else :
                    r = mid
            return l
        left = binSeach(target)
        right = binSeach(target+1) - 1
        if left == len(nums) or nums[left] != target: #做特判不等于len,即tgt不大于所有数
            return [-1, -1]
        return [left,right]
if __name__ ==  "__main__":
    solution = Solution()
    nums = [1,2,3]
    target = 4
    print(solution.searchRange(nums,target))

File: code/code34.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        if (n <= 0) return {-1,-1};
        int l = 0,r = n-1;
        int left = binSeach(nums,target-1);
        int right = binSeach(nums,target) - 1;
        if (left >= n||nums[left] != target) {
            return {-1,-1};
        }
        return {left,right};        
    }
    int binSeach(vector<int>& nums,int target) {
        int n = nums.size();
        int l = 0,r = n-1;
        while (l <= r) {
            int mid = l + r >> 1;
            if (nums[mid] > target) {
                r = mid - 1;
            }else {
                l = mid + 1;
            }
        }
        return l;
    }
};
int main() {
    Solution solution;
    vector<int> nums = {5,7,7,8,8,10};
    vector<int> ans = solution.searchRange(nums,8);
    cout << ans[0] <<" "<<  ans[1] << endl;
}